libraries/AP_RangeFinder/AP_RangeFinder_Benewake_TFS20L.cpp

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (C) 2024
+ *
+ * This file is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This file is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "AP_RangeFinder_Benewake_TFS20L.h"
+
+#if AP_RANGEFINDER_BENEWAKE_TFS20L_ENABLED
+
+#include <utility>
+
+#include <GCS_MAVLink/GCS.h>
+#include <AP_HAL/AP_HAL.h>
+#include <AP_HAL/I2CDevice.h>
+
+extern const AP_HAL::HAL& hal;
+
+#define DRIVER "TFS20L"
+#define BENEWAKE_OUT_OF_RANGE_ADD_CM 100
+
+
+// TFS20L register addresses
+static const uint8_t TFS20L_DIST_LOW = 0x00;         // Distance low byte
+static const uint8_t TFS20L_DIST_HIGH = 0x01;        // Distance high byte  
+static const uint8_t TFS20L_AMP_LOW = 0x02;          // Signal strength/amplitude low byte
+static const uint8_t TFS20L_AMP_HIGH = 0x03;         // Signal strength/amplitude high byte
+static const uint8_t TFS20L_TEMP_LOW = 0x04;         // Temperature low byte
+static const uint8_t TFS20L_TEMP_HIGH = 0x05;        // Temperature high byte
+static const uint8_t TFS20L_VERSION_REVISION = 0x0A; // Version revision byte
+static const uint8_t TFS20L_VERSION_MINOR = 0x0B;    // Version minor byte
+static const uint8_t TFS20L_VERSION_MAJOR = 0x0C;    // Version major byte
+static const uint8_t TFS20L_ENABLE = 0x25;           // Enable register
+
+// Distance and strength limits
+static const uint16_t MAX_DIST_CM = 2000;
+static const uint16_t MIN_DIST_CM = 1;
+static const uint16_t MIN_STRENGTH = 100;
+
+
+AP_RangeFinder_Benewake_TFS20L::AP_RangeFinder_Benewake_TFS20L(
+        RangeFinder::RangeFinder_State &_state,
+        AP_RangeFinder_Params &_params,
+        AP_HAL::I2CDevice *dev)
+    : AP_RangeFinder_Backend(_state, _params)
+    , _dev(dev)
+{
+}
+
+AP_RangeFinder_Backend *AP_RangeFinder_Benewake_TFS20L::detect(
+        RangeFinder::RangeFinder_State &_state,
+        AP_RangeFinder_Params &_params,
+        AP_HAL::I2CDevice *dev)
+{
+    if (dev == nullptr) {
+        return nullptr;
+    }
+    AP_RangeFinder_Benewake_TFS20L *sensor = NEW_NOTHROW AP_RangeFinder_Benewake_TFS20L(_state, _params, dev);
+    if (sensor == nullptr || !sensor->init()) {
+        delete sensor;
+        return nullptr;
+    }
+    return sensor;
+}
+
+bool AP_RangeFinder_Benewake_TFS20L::init()
+{
+    uint8_t version_data[3];
+
+    {
```

Comment:
Nice.

I think we decided a while back it was safe to hold this semaphore across the registration of the periodic callback.